waspc/test/AppSpec/ValidTest.hs

```diff
@@ -380,17 +380,86 @@ spec_AppSpecValid = do
+
+    describe "Prisma schema validation" $ do
+      it "should validate that Prisma models are not using unsupported field modifiers" $ do
+        let prismaSchema =
+              Psl.Ast.Schema $
+                minimalPassingPrismaConfig
+                  ++ [ Psl.Ast.SchemaModel $
+                         Psl.Ast.Model "User" $
+                           Psl.Ast.Body
+                             [ Psl.Ast.ElementField $
+                                 Psl.Ast.Field
+                                   { Psl.Ast._name = "id",
+                                     Psl.Ast._type = Psl.Ast.String,
+                                     Psl.Ast._typeModifiers =
+                                       [ Psl.Ast.UnsupportedOptionalList
+                                       ],
+                                     Psl.Ast._attrs = []
+                                   }
+                             ]
+                     ]
+        let appSpec = basicAppSpec {AS.prismaSchema = prismaSchema}
+        ASV.validateAppSpec appSpec `shouldBe` [ASV.GenericValidationError "Model \"User\" in schema.prisma has defined \"id\" field as an optional list, which is not supported by Prisma."]
+
+    it "should validate that some datasource exists" $
+      let prismaSchema =
+            Psl.Ast.Schema
+              [ Psl.Ast.SchemaGenerator $
+                  Psl.Ast.Generator
+                    "client"
+                    [ Psl.Ast.ConfigBlockKeyValue "provider" "\"prisma-client-js\""
+                    ]
+              ]
+       in ASV.validateAppSpec (basicAppSpec {AS.prismaSchema = prismaSchema})
+            `shouldBe` [ASV.GenericValidationError "Prisma schema must have exactly one datasource defined."]
+
+    it "should validate that only one datasource exists" $
+      let prismaSchema =
+            Psl.Ast.Schema
+              [ Psl.Ast.SchemaDatasource $ Psl.Ast.Datasource "db" [],
+                Psl.Ast.SchemaDatasource $ Psl.Ast.Datasource "db" [],
+                Psl.Ast.SchemaGenerator $
+                  Psl.Ast.Generator
+                    "client"
+                    [ Psl.Ast.ConfigBlockKeyValue "provider" "\"prisma-client-js\""
+                    ]
+              ]
+       in ASV.validateAppSpec (basicAppSpec {AS.prismaSchema = prismaSchema})
+            `shouldBe` [ASV.GenericValidationError "Prisma schema must have exactly one datasource defined."]
+
+    it "should validate that there is at least one generator" $
+      let prismaSchema =
+            Psl.Ast.Schema
+              [ Psl.Ast.SchemaDatasource $ Psl.Ast.Datasource "db" []
+              ]
+       in ASV.validateAppSpec (basicAppSpec {AS.prismaSchema = prismaSchema})
+            `shouldBe` [ASV.GenericValidationError "Prisma schema should have at least one generator defined."]
+
+    it "should validate that there is at least one client generator" $
+      let prismaSchema =
```

Comment:
It feels a bit more complex written like this: 

```hs
    describe "Prisma schema validation" $ do
      it "should validate that Prisma models are not using unsupported field modifiers" $ do
        let prismaFile =
              T.unpack
                [trimming|
                datasource db {}

                generator client {
                  provider = "prisma-client-js"
                }

                model User {
                  id String[]?
                }
                |]
        case Psl.Parser.parsePrismaSchema prismaFile of
          Right prismaSchema -> do
            let appSpec = basicAppSpec {AS.prismaSchema = prismaSchema}
            ASV.validateAppSpec appSpec `shouldBe` [ASV.GenericValidationError "Model \"User\" in schema.prisma has defined \"id\" field as an optional list, which is not supported by Prisma."]
          Left _ -> pure ()
```

Since I have to take into account that the schema parsing might fail, I have to handle `Either` each time.

Reply:
Hm right, that casing is not fun.
But what you can do is put a `fromRight` in front of that `parsePrismaSchema` and just give it `error` or something as a function (or value) in case of failure.

Reply:
I've written a test helper and it simplified down to this:
```hs
let prismaSchema =
  Util.getPrismaSchema
    [trimming|
      datasource db {}
    |]
```